### ISSUE TITLE
Missed heartbeat caused permanent cluster inconsistency

### ISSUE THREAD

**Issue by torkeld** [hazelcast/hazelcast#6108]
*I apologize for the long description.*

TL;DR: This happened in our production environment a week ago. There was a short network outage that caused a missed heartbeat between the master node and one of the members. That triggered a series of events that eventually left the cluster in a state where the different nodes had different masters and different members list. The result was operations that failed with CallerNotMemberException and WrongTargetException.

Hazelcast version used: 3.2.6

#### Nodes
CS1: IP 10.181.123.80, AZ 1b
CS2: IP 10.113.170.130, AZ 1e
CS3: IP 10.232.137.233, AZ 1d, initially master
CS4: IP 10.166.134.204, AZ 1d

#### Sequence of events
![analysis of incident 86685349888](https://cloud.githubusercontent.com/assets/6555813/9610112/608800da-508c-11e5-8892-319c72d2f3f7.jpg)

#### Analysis
Hazelcast nodes send heartbeats every 5s. If a node doesn’t receive a heartbeat within 10s it will remove that node from the cluster.
In this scenario, CS3, who is the master of the cluster, doesn’t get a heartbeat from CS2 in 10s so it removes it from the cluster. CS2 on the other hand does get the heartbeat from CS3 so it doesn’t remove CS3. At this point CS2 still has all the other members in its cluster state, but the other members don’t have CS2. Now, the other members try to reconnect to CS2 (see SplitBrainHandler.java). CS3 sees that CS2 has more members than itself (4 vs 3) it decides to merge with CS2.

When CS3 tries to join CS2’s cluster, CS2 sees that CS3 is still in its member list so it first removes it from its cluster. Since CS2 was master CS3 will assign a new master, CS4.
```
WARN  20:43:28.539  [10.113.170.130]:5701 [prod-voice-conference-state-service] [3.2.6] New join request has been received from an existing endpoint!
 => Member [10.232.137.233]:5701 Removing old member and processing join request...
INFO  20:43:28.539  [10.113.170.130]:5701 [prod-voice-conference-state-service] [3.2.6] Master Address[10.232.137.233]:5701 left the cluster. Assigning new master Member [10.166.134.204]:5701
```
CS2 tells CS3 that CS4 is the new master, and when CS4 joins CS2 does the same thing with CS4 and now promotes itself to master.
```
WARN  20:43:28.630 [10.113.170.130]:5701 [prod-voice-conference-state-service] [3.2.6] New join request has been received from an existing endpoint! => Member [10.166.134.204]:5701 Removing old member and processing join request...
INFO  20:43:28.630 [10.113.170.130]:5701 [prod-voice-conference-state-service] [3.2.6] Master Address[10.166.134.204]:5701 left the cluster. Assigning new master Member [10.113.170.130]:5701 this
```
For some reason CS3 is not informed that CS2 is now master and ignores the master responses from CS2 as it thinks CS4 is master.
```
WARN  20:43:31.540 [10.232.137.233]:5701 [prod-voice-conference-state-service] [3.2.6] Ignoring master response from Address[10.113.170.130]:5701, since this node has an active master: Address[10.166.134.204]:5701

Similarly, CS1 didn’t get notified when master changed and also ignores the master responses from CS2 as it thinks CS3 is master.
WARN  20:43:31.521 [10.181.123.80]:5701 [prod-voice-conference-state-service] [3.2.6] Ignoring master response from Address[10.113.170.130]:5701, since this node has an active master: Address[10.232.137.233]:5701
```
CS2 seems to get stuck merging at this point and keeps printing the following message, while all other members are done with the merge.
```
INFO  20:43:37.777 [10.113.170.130]:5701 [prod-voice-conference-state-service] [3.2.6] Re-partitioning cluster data... Migration queue size: 203
INFO  20:43:43.959 [10.113.170.130]:5701 [prod-voice-conference-state-service] [3.2.6] Remaining migration tasks in queue => 134
INFO  20:43:58.959 [10.113.170.130]:5701 [prod-voice-conference-state-service] [3.2.6] Remaining migration tasks in queue => 134
INFO  20:44:13.959 [10.113.170.130]:5701 [prod-voice-conference-state-service] [3.2.6] Remaining migration tasks in queue => 134
This is probably because CS3 doesn’t think that CS2 is the master and ignores the migration tasks.
WARN  20:43:51.437 [10.232.137.233]:5701 [prod-voice-conference-state-service] [3.2.6] Address[10.181.123.80]:5701 has sent MasterConfirmation, but this node is not master!
```
At this point all nodes have 4 nodes in their member lists but a different view of who is master. HTTP requests at this point are being served properly (although there might have been some data inconsistency), entry processors are being processed by all nodes.

About 8 minutes later (500 s) CS2 removes CS1 from the cluster since it hasn’t received any master confirmation in 500 seconds (CS1 has been sending those to CS3 the whole time). It instructs CS3 and CS4 to remove CS1, but CS3 ignores this since it thinks that CS4 is master.
```
WARN  20:51:48.867 [10.232.137.233]:5701 [prod-voice-conference-state-service] [3.2.6] Received a ClusterRuntimeState, but its sender doesn't seem to be master! => Sender: Address[10.113.170.130]:5701, Master: Address[10.166.134.204]:5701! (Ignore if master node has changed recently.)
```
2s later CS2 reconnects with CS1 and since CS1 has more members in its list (4 vs 3) CS2 will merge with CS1, and hence CS2 will now consider CS3 as master (same as CS1). However, CS2 doesn’t add CS1 to its members list. 
```
INFO  20:52:00.822 [10.113.170.130]:5701 [prod-voice-conference-state-service] [3.2.6] Address[10.113.170.130]:5701 is merging to Address[10.181.123.80]:5701, because : joinRequest.getMemberCount() > currentMemberCount [4 > 3]
```
CS4 also merges with CS1 as instructed by CS2.
```
WARN  20:52:00.827 [10.166.134.204]:5701 [prod-voice-conference-state-service] [3.2.6] Address[10.166.134.204]:5701 is merging to Address[10.181.123.80]:5701, because: instructed by master Address[10.113.170.130]:5701
```
CS4 “re-joins” with CS3. This causes CS3 to remove CS4 and assigning CS2 as master:
```
WARN  20:52:02.946 [10.232.137.233]:5701 [prod-voice-conference-state-service] [3.2.6] New join request has been received from an existing endpoint! => Member [10.166.134.204]:5701 Removing old member and processing join request...
ERROR 20:52:02.946 [10.232.137.233]:5701 [prod-voice-conference-state-service] [3.2.6] Old master Address[10.166.134.204]:5701 is dead but the first of member list is a different member Member [10.113.170.130]:5701!
INFO  20:52:02.946 [10.232.137.233]:5701 [prod-voice-conference-state-service] [3.2.6] Master Address[10.166.134.204]:5701 left the cluster. Assigning new master Member [10.113.170.130]:5701
INFO  20:52:02.946 [10.232.137.233]:5701 [prod-voice-conference-state-service] [3.2.6] Removing Member [10.166.134.204]:5701
```
Similarly, CS2 re-joins with CS3, which causes CS3 to remove CS2 and assigning CS1 as master.
```
WARN  20:52:02.998 [10.232.137.233]:5701 [prod-voice-conference-state-service] [3.2.6] New join request has been received from an existing endpoint! => Member [10.113.170.130]:5701 Removing old member and processing join request...
INFO  20:52:02.998 [10.232.137.233]:5701 [prod-voice-conference-state-service] [3.2.6] Master Address[10.113.170.130]:5701 left the cluster. Assigning new master Member [10.181.123.80]:5701
INFO  20:52:02.998 [10.232.137.233]:5701 [prod-voice-conference-state-service] [3.2.6] Removing Member [10.113.170.130]:5701
```
CS1 now thinks CS2 is master and starts sending master responses to it (why??):
```
WARN  20:52:03.008 [10.113.170.130]:5701 [prod-voice-conference-state-service] [3.2.6] Ignoring master response from Address[10.181.123.80]:5701, since this node has an active master: Address[10.232.137.233]:5701
```
From this point HTTP traffic starts to fail since CS3 and CS4 rejects all entry processors from CS1 as they are not in their members lists.

CS2 and CS4 keeps trying to rejoin with CS3 but fails:
```
WARN  20:54:39.023 [10.113
.170.130]:5701 [prod-voice-conference-state-service] [3.2.6] Couldn't join to the master : Address[10.232.137.233]:5701
WARN  20:54:39.023 [10.113
.170.130]:5701 [prod-voice-conference-state-service] [3.2.6] Failed to connect, node joined= false, allConnected= false to all other me
mbers after 0 seconds.
WARN  20:54:39.023 [10.113
.170.130]:5701 [prod-voice-conference-state-service] [3.2.6] Rebooting after 10 seconds.
```
And eventually gives up and sets itself as master:
```
WARN  20:55:16.029 [10.113.170.130]:5701 [prod-voice-conference-state-service] [3.2.6] Join try count exceed limit, setting this node as master!
```

**Comment by torkeld** [comment]
Full logs can be provided upon request.

**Comment by enesakar** [comment]
Hi @torkeId 
There are some changes (improvements) on heartbeats and master selection on 3.5.
Is it possible for you to reproduce the issue with 3.5.2?

**Comment by torkeld** [comment]
We've tried to reproduce this in a test environment on both 3.2.6 and 3.5.2 without luck. It's likely very timing sensitive, but was hoping that you can make some sense out of it and with some level confidence tell whether it is fixed in 3.5.2.

Thanks,
Torkel (@Twilio)

**Comment by tombujok** [comment]
Thanks for your feedback. We're looking at it right now. Could share full logs somewhere (dropbox?)

**Comment by torkeld** [comment]
https://app.box.com/s/ku9wbgonpe076c7xik4wurwxgud7nhe5

Thanks!

**Comment by tombujok** [comment]
Thx for the logs. We are looking into it, but it may take us some time to make any statement, so plz state tuned.

**Comment by tombujok** [comment]
Thanks for your patience. We have analysed your ticket and thanks to an improvement that has been added to version 3.3, Hazelcast would behave correctly in a half-split scenario, that has been described by you in the ticket. 

You can track the fix by looking at the highlighted code snippets:

* Old version: https://github.com/hazelcast/hazelcast/blob/v3.2.7/hazelcast/src/main/java/com/hazelcast/cluster/ClusterServiceImpl.java#L528
* New version: https://github.com/hazelcast/hazelcast/blob/v3.3/hazelcast/src/main/java/com/hazelcast/cluster/ClusterServiceImpl.java#L563


In the meanwhile we would also advise you to increase the MAX_NO_HEARTBEAT_SECONDS to 30 seconds. If we understand your config correctly, it looks as follows:
`HEARTBEAT_INTERVAL_SECONDS == 5`
`MAX_NO_HEARTBEAT_SECONDS == 10`

It's risky, since two dropped heartbeats may invoke the Master election & Join mechanisms, which have their own momentum, so it would be advisable to make sure that it doesn't happen on short-term connectivity problems.

**Comment by torkeld** [comment]
Hi Tomek,

Thanks for taking your time to look into this!
Can I assume that the same fix is in 3.5.2? In that case we'll go ahead and upgrade directly to 3.5.2, which we have tested for some time with good result.
Would you also recommend bumping the MAX_NO_HEARTBEAT_SECONDS to 30 when on version 3.5.2?

Regards,
Torkel

**Comment by tombujok** [comment]
Hi Torkel,
Yes, it's fixed in 3.5.2. too, so you are good to go with the upgrade.
Yes, I would recommend bumping the MAX_NO_HEARTBEAT_SECONDS to 30 seconds. 
It's reasonable to make the difference between `HEARTBEAT_INTERVAL_SECONDS` and `MAX_NO_HEARTBEAT_SECONDS` bigger.
